engine/crates/composition/src/compose/context.rs

```diff
@@ -241,6 +225,28 @@ impl<'a> Context<'a> {
             None => self.ir.insert_static_str(string),
         }
     }
+
+    pub(crate) fn set_query(&mut self, id: federated::ObjectId) {
+        self.ir.query_type = Some(id);
+    }
+
+    pub(crate) fn set_mutation(&mut self, id: federated::ObjectId) {
+        self.ir.mutation_type = Some(id);
+    }
+
+    pub(crate) fn set_subscription(&mut self, id: federated::ObjectId) {
+        self.ir.subscription_type = Some(id);
+    }
+
+    pub(crate) fn insert_object_field_from_entity_interface(
```

Comment:
This had to be split as a cascading change, because of string interning constants. I find the new implementation more straightforward regardless.